File: backend/app/models/order.py
"""
Models: ORDERS, ORDER_SHIPMENTS, ORDER_ITEMS

Referências:
- ARCHITECTURE.md §ERD: ORDERS, ORDER_SHIPMENTS, ORDER_ITEMS
- ARCHITECTURE.md §Notas de Modelagem:
  - ORDER_SHIPMENTS viabiliza split multi-loja (frete, status e SLA próprios)
  - ORDER_ITEMS aponta para shipment_id (separação logística sem duplicar pedido)
  - ORDERS.idempotency_key garante idempotência em criação de pedidos
- ARCHITECTURE.md §Checkout Crítico: fluxo completo de criação com stock lock
"""

import uuid
from datetime import datetime

from sqlalchemy import (
    String, DateTime, Float, Integer, ForeignKey,
    Enum as SAEnum, Text, Numeric,
)
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.database import Base

import enum


class OrderStatus(str, enum.Enum):
    PENDING = "pending"
    CONFIRMED = "confirmed"
    PREPARING = "preparing"
    READY_FOR_PICKUP = "ready_for_pickup"
    OUT_FOR_DELIVERY = "out_for_delivery"
    DELIVERED = "delivered"
    CANCELLED = "cancelled"
    REFUNDED = "refunded"


class PaymentStatus(str, enum.Enum):
    PENDING = "pending"
    PROCESSING = "processing"
    PAID = "paid"
    FAILED = "failed"
    REFUNDED = "refunded"


class ShipmentStatus(str, enum.Enum):
    PENDING = "pending"
    CONFIRMED = "confirmed"
    PREPARING = "preparing"
    SHIPPED = "shipped"
    DELIVERED = "delivered"
    CANCELLED = "cancelled"


class Order(Base):
    """
    ORDERS: pedido principal do cliente. Contém idempotency_key para evitar duplicidades
    (ARCHITECTURE.md §Headers Obrigatórios).
    """
    __tablename__ = "orders"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    user_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("users.id"), nullable=False, index=True
    )
    status: Mapped[OrderStatus] = mapped_column(SAEnum(OrderStatus), default=OrderStatus.PENDING)
    payment_status: Mapped[PaymentStatus] = mapped_column(
        SAEnum(PaymentStatus), default=PaymentStatus.PENDING
    )
    subtotal: Mapped[float] = mapped_column(Numeric(10, 2), nullable=False)
    delivery_fee_total: Mapped[float] = mapped_column(Numeric(10, 2), default=0)
    discount: Mapped[float] = mapped_column(Numeric(10, 2), default=0)
    total: Mapped[float] = mapped_column(Numeric(10, 2), nullable=False)
    coupon_code: Mapped[str | None] = mapped_column(String(50), nullable=True)
    notes: Mapped[str | None] = mapped_column(Text, nullable=True)
    delivery_address_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), ForeignKey("addresses.id"), nullable=True
    )
    # Idempotência (ARCHITECTURE.md §Contratos de API)
    idempotency_key: Mapped[str] = mapped_column(String(64), unique=True, nullable=False, index=True)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=datetime.utcnow)
    updated_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), onupdate=datetime.utcnow)

    # Relationships
    user = relationship("User", back_populates="orders")
    shipments = relationship("OrderShipment", back_populates="order", lazy="selectin")
    payments = relationship("Payment", back_populates="order", lazy="selectin")
    rma_requests = relationship("RmaRequest", back_populates="order", lazy="selectin")


class OrderShipment(Base):
    """
    ORDER_SHIPMENTS: split de pedido por loja.
    Cada shipment possui frete, status e SLA independentes
    (ARCHITECTURE.md §Notas de Modelagem).
    """
    __tablename__ = "order_shipments"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    order_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("orders.id"), nullable=False, index=True
    )
    store_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("stores.id"), nullable=False, index=True
    )
    status: Mapped[ShipmentStatus] = mapped_column(
        SAEnum(ShipmentStatus), default=ShipmentStatus.PENDING
    )
    delivery_fee: Mapped[float] = mapped_column(Numeric(10, 2), default=0)
    estimated_delivery: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=datetime.utcnow)

    # Relationships
    order = relationship("Order", back_populates="shipments")
    store = relationship("Store", back_populates="shipments")
    items = relationship("OrderItem", back_populates="shipment", lazy="selectin")


class OrderItem(Base):
    """
    ORDER_ITEMS: vinculado a shipment_id para separação logística
    (ARCHITECTURE.md §Notas de Modelagem).
    """
    __tablename__ = "order_items"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    shipment_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("order_shipments.id"), nullable=False, index=True
    )
    product_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("products.id"), nullable=False
    )
    sku_variant_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("sku_variants.id"), nullable=False
    )
    quantity: Mapped[int] = mapped_column(Integer, nullable=False)
    unit_price: Mapped[float] = mapped_column(Numeric(10, 2), nullable=False)
    total_price: Mapped[float] = mapped_column(Numeric(10, 2), nullable=False)

    # Relationships
    shipment = relationship("OrderShipment", back_populates="items")
    rma_items = relationship("RmaItem", back_populates="order_item", lazy="selectin")

File: apps/zoe_app/lib/domain/entities/order.dart
/// Entidade Order do domínio.
///
/// Referência: ARCHITECTURE.md §ERD: ORDERS, ORDER_SHIPMENTS
class Order {
  final String id;
  final String userId;
  final String status;
  final double subtotal;
  final double deliveryFee;
  final double discount;
  final double total;
  final String? couponCode;
  final String deliveryAddressId;
  final DateTime createdAt;
  final List<OrderShipment> shipments;

  const Order({
    required this.id,
    required this.userId,
    required this.status,
    required this.subtotal,
    required this.deliveryFee,
    this.discount = 0,
    required this.total,
    this.couponCode,
    required this.deliveryAddressId,
    required this.createdAt,
    required this.shipments,
  });

  factory Order.fromJson(Map<String, dynamic> json) {
    return Order(
      id: json['id'] as String,
      userId: json['user_id'] as String,
      status: json['status'] as String,
      subtotal: (json['subtotal'] as num).toDouble(),
      deliveryFee: (json['delivery_fee'] as num?)?.toDouble() ?? 0,
      discount: (json['discount_amount'] as num?)?.toDouble() ?? 0,
      total: (json['total_amount'] as num).toDouble(),
      couponCode: json['coupon_code'] as String?,
      deliveryAddressId: json['delivery_address_id'] as String? ?? '',
      createdAt: DateTime.parse(json['created_at'] as String),
      shipments: (json['shipments'] as List<dynamic>?)
              ?.map((e) => OrderShipment.fromJson(e as Map<String, dynamic>))
              .toList() ??
          [],
    );
  }

  Map<String, dynamic> toJson() => {
        'id': id,
        'user_id': userId,
        'status': status,
        'subtotal': subtotal,
        'delivery_fee': deliveryFee,
        'discount_amount': discount,
        'total_amount': total,
        'coupon_code': couponCode,
        'delivery_address_id': deliveryAddressId,
        'created_at': createdAt.toIso8601String(),
        'shipments': shipments.map((s) => s.toJson()).toList(),
      };
}

class OrderShipment {
  final String id;
  final String orderId;
  final String storeId;
  final String storeName;
  final String status;
  final String? trackingCode;
  final List<OrderItem> items;

  const OrderShipment({
    required this.id,
    required this.orderId,
    required this.storeId,
    required this.storeName,
    required this.status,
    this.trackingCode,
    required this.items,
  });

  factory OrderShipment.fromJson(Map<String, dynamic> json) {
    return OrderShipment(
      id: json['id'] as String,
      orderId: json['order_id'] as String? ?? '',
      storeId: json['store_id'] as String,
      storeName: json['store_name'] as String? ?? '',
      status: json['status'] as String,
      trackingCode: json['tracking_code'] as String?,
      items: (json['items'] as List<dynamic>?)
              ?.map((e) => OrderItem.fromJson(e as Map<String, dynamic>))
              .toList() ??
          [],
    );
  }

  Map<String, dynamic> toJson() => {
        'id': id,
        'order_id': orderId,
        'store_id': storeId,
        'store_name': storeName,
        'status': status,
        'tracking_code': trackingCode,
        'items': items.map((i) => i.toJson()).toList(),
      };
}

class OrderItem {
  final String id;
  final String productId;
  final String productName;
  final String variantId;
  final String size;
  final String color;
  final int quantity;
  final double unitPrice;
  final String? imageUrl;

  const OrderItem({
    required this.id,
    required this.productId,
    required this.productName,
    required this.variantId,
    required this.size,
    required this.color,
    required this.quantity,
    required this.unitPrice,
    this.imageUrl,
  });

  factory OrderItem.fromJson(Map<String, dynamic> json) {
    return OrderItem(
      id: json['id'] as String,
      productId: json['product_id'] as String? ?? '',
      productName: json['product_name'] as String? ?? '',
      variantId: json['variant_id'] as String? ?? json['sku_variant_id'] as String? ?? '',
      size: json['size'] as String? ?? '',
      color: json['color'] as String? ?? '',
      quantity: json['quantity'] as int,
      unitPrice: (json['unit_price'] as num).toDouble(),
      imageUrl: json['image_url'] as String?,
    );
  }

  Map<String, dynamic> toJson() => {
        'id': id,
        'product_id': productId,
        'product_name': productName,
        'variant_id': variantId,
        'size': size,
        'color': color,
        'quantity': quantity,
        'unit_price': unitPrice,
        'image_url': imageUrl,
      };
}
